File: backend/src/schemas/ws_events.py
"""WebSocket event protocol schemas."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel

WSEventType = Literal[
    "agent_start",
    "agent_end",
    "thinking",
    "stream_start",
    "stream_token",
    "stream_end",
    "progress",
    "approval_gate",
    "error",
]


class WSEvent(BaseModel):
    """A single WebSocket event sent from server to client."""

    type: WSEventType
    agent: str | None = None
    message: str | None = None
    section: str | None = None
    token: str | None = None
    current: int | None = None
    total: int | None = None
    label: str | None = None
    gate: str | None = None
    data: dict[str, Any] | None = None
    recoverable: bool | None = None

File: frontend/src/types/ws_events.ts
export type WSEventType =
  | "agent_start"
  | "agent_end"
  | "thinking"
  | "stream_start"
  | "stream_token"
  | "stream_end"
  | "progress"
  | "approval_gate"
  | "error";

export interface WSEvent {
  type: WSEventType;
  agent?: string | null;
  message?: string | null;
  section?: string | null;
  token?: string | null;
  current?: number | null;
  total?: number | null;
  label?: string | null;
  gate?: string | null;
  data?: Record<string, unknown> | null;
  recoverable?: boolean | null;
}

export interface Progress {
  current: number;
  total: number;
  label: string | null;
}
